### PR TITLE
Add checking the terminal is xterm compatible

### PR DESCRIPTION
**Pull request by ttdoda**
Some terminals do not properly handle DCS string and/or CSI sequence with intermediate byte.
Therefore, when get the cursor style, garbage string is displayed.

This patch detects such a terminal, checking with termresponse becomes unnecessary.
This patch solves most case of #2008

Tested with following terminals:
  * xterm pl#330
  * mlterm 3.8.2
  * rxvt 2.6.4
  * rxvt-unicode 9.22
  * xvt 3.0.0
  * Gnome Terminal and other VTE-based terminals (VTE 0.44.2, 0.42.4, 0.28.2)
  * Konsole 16.12.3
  * Terminal.app (macOS 10.12.6)
  * iTerm2
  * Tera Term 4.96
  * PuTTY 0.70
  * SecureCRT 6.7.4(\*), 7.0.0(\*), 8.1.4
  * Poderosa 4.4.1(\*), 5.4.2(\*)
  * Absolute Telnet 9.53(*), 10.16(\*)
  * TTYEmulator 4.3.0.2
  * GNU Screen 4.03.01
  * tmux 2.5

(\*) has problem with plain vim 8.0.1127.

### CODE REVIEW

**Comment by nuko8** [comment]
It could be very helpful for the review if you would add your own comment on what will happen to Vim working over SSH of a various combination of platforms and terminals after the proposed patch is merged.

**Comment by brammool** [comment]
Kouichi Iwamoto wrote:

> Some terminals do not properly handle DCS string and/or CSI sequence
> with intermediate byte.
> Therefore, when get the cursor style, garbage string is displayed.
> 
> This patch detects such a terminal, checking with termresponse becomes
> unnecessary.
> This patch solves most case of #2008
> 
> Tested with following terminals:
>   * xterm pl#330
>   * mlterm 3.8.2
>   * rxvt 2.6.4
>   * rxvt-unicode 9.22
>   * xvt 3.0.0
>   * Gnome Terminal and other VTE-based terminals (VTE 0.44.2, 0.42.4, 0.28.2)
>   * Konsole 16.12.3
>   * Terminal.app (macOS 10.12.6)
>   * iTerm2
>   * Tera Term 4.96
>   * PuTTY 0.70
>   * SecureCRT 6.7.4(\*), 7.0.0(\*), 8.1.4
>   * Poderosa 4.4.1(\*), 5.4.2(\*)
>   * Absolute Telnet 9.53(*), 10.16(\*)
>   * TTYEmulator 4.3.0.2
>   * GNU Screen 4.03.01
>   * tmux 2.5
> 
> (\*) has problem with plain vim 8.0.1127.

That's an interesting idea.
There were some problems with the similar method used to obtain the
ambiguous width, but I think they were all solved.

Can we integrate the two checks, to reduce the number of flushes and
especially the vpeek calls?  Also avoid the cursor moving around
multiple times.

-- 
hundred-and-one symptoms of being an internet addict:
166. You have been on your computer soo long that you didn't realize
     you had grandchildren.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
@brammool Do you mean that we should make new function that do both of may_req_ambiguous_char_width and may_req_xterm_compat_test? Or simply reducing out_flush in may_req_xterm_compat_test? The first one may be bit confusable to read code.

**Comment by nuko8** [comment]
> There were some problems with the similar method used to obtain the ambiguous width, but I think they were all solved.

Really?  *shrug*.   I'm wondering why this sort of  problems is persistent with Vim in comparison with other pieces of OSS software.   Seems like this community prefers louder voices. and decisions are made more politically than technically, but people hardly know its politics.

Anyway, while  I like the idea, but not-mentioning latency makes me uneasy.  Is this trivial?  Then just mention it; otherwise, we will need another tweak to make our test suite not-flaky.

**Comment by brammool** [comment]
Yasuhiro wrote:

> @brammool Do you mean that we should make new function that do both of
> may_req_ambiguous_char_width and may_req_xterm_compat_test? Or simply
> reducing out_flush in may_req_xterm_compat_test? The first one may be
> bit confusable to read code.

Make it one function.  Yes, it will be a bit harder to read, but
reducing the overhead is relevant.  Especially over remote connections.
And once it works we don't have to read it again :-).

-- 
"The question of whether computers can think is just like the question
of whether submarines can swim."      -- Edsger W. Dijkstra

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ttdoda** [comment]
> Make it one function. 

updated the patch. Could you check it?

**Comment by k-takata** [comment]
Could you resolve conflicts?

**Comment by k-takata** [comment]
@ttdoda Hmm, it looks you did something wrong. The history looks broken on the PR page.
I think the following commands fix the problem:

```
$ git checkout check-dcs-handling
$ git reset --hard 9cfc6cb
$ git push -f ttdoda HEAD  # Change "ttdoda" to appropriate remote name if needed. 
```

File: src/term.c [hunk 1]
```diff
@@ -4511,6 +4555,15 @@ check_termcode(
 # endif
 			}
 		    }
+		    /* Third row is xterm compatibility test */
+		    else if (row_char == '3') {
+			if (col != 1) {
```
Comment (k-takata) [inline src/term.c]: Could you move the last `{` to the next line? (also for the above line.)

Comment (ttdoda) [reply src/term.c]: updated. thanks!

**Comment by k-takata** [comment]
@ttdoda Thank you for updating.

@brammool I don't know why you lower the priority of this PR in the todo.txt, but I think this is ready to be merged.

**Comment by k-takata** [comment]
Oh, 8.0.1303 causes conflict again. @ttdoda, could you resolve it?

**Comment by k-takata** [comment]
@ttdoda Thank you. AppVeyor build failed, but I don't think it is caused by your patch.
I tried your patch on my account, and it successfully passed:
https://ci.appveyor.com/project/k-takata/vim/build/170

**Comment by brammool** [comment]
This is tricky, but since we keep having problems with various misbehaving terminals I think we should include this.  The alternative is to build out the version response checking, which is tedious but perhaps a better long-term solution.  We'll see how it goes.

**Pull request closed without merging**